addons/hr_holidays/models/hr_leave.py

```diff
@@ -891,10 +891,71 @@ def action_validate(self):
             else:
                 employees = holiday.department_id.member_ids
 
-            if self.env['hr.leave'].search_count([('date_from', '<=', holiday.date_to), ('date_to', '>', holiday.date_from),
-                               ('state', 'not in', ['cancel', 'refuse']), ('holiday_type', '=', 'employee'),
-                               ('employee_id', 'in', employees.ids)]):
-                raise ValidationError(_('You can not have 2 leaves that overlaps on the same day.'))
+            conflicting_leaves = self.env['hr.leave'].with_context(
+                tracking_disable=True,
+                mail_activity_automation_skip=True,
+                leave_fast_create=True
+            ).search([
+                ('date_from', '<=', holiday.date_to),
+                ('date_to', '>', holiday.date_from),
+                ('state', 'not in', ['cancel', 'refuse']),
+                ('holiday_type', '=', 'employee'),
+                ('employee_id', 'in', employees.ids)])
+
+            if conflicting_leaves:
+                # YTI: More complex use cases could be managed in master
+                if holiday.leave_type_request_unit != 'day' or any(l.leave_type_request_unit == 'hour' for l in conflicting_leaves):
+                    raise ValidationError(_('You can not have 2 leaves that overlaps on the same day.'))
+
+                for conflicting_leave in conflicting_leaves:
+                    if conflicting_leave.leave_type_request_unit == 'half_day' and conflicting_leave.request_unit_half:
+                        conflicting_leave.action_refuse()
+                        continue
+                    # Leaves in days
+                    split_leaves = self.env['hr.leave']
+                    target_state = conflicting_leave.state
+                    conflicting_leave.action_refuse()
+                    if conflicting_leave.date_from < holiday.date_from:
+                        before_leave_vals = conflicting_leave.copy_data({
+                            'date_from': conflicting_leave.date_from.date(),
+                            'date_to': holiday.date_from.date() + timedelta(days=-1),
+                        })[0]
+                        before_leave = self.env['hr.leave'].new(before_leave_vals)
+                        before_leave._onchange_request_parameters()
+                        # Could happen for part-time contract, that time off is not necessary
+                        # anymore.
+                        if before_leave.date_from < before_leave.date_to:
```

Comment:
I don't understand the link between this `if` statement and part-time contracts.

Reply:
Imagine you work on monday-wednesday-friday only.

You take a time off on friday.

We create a company time off on friday.

By looking at the last attendance before the company time off start date to compute the date_to, you would have a date_from > date_to. Just don't create the leave at that time.

Reply:
haan oui, because the leave is not actually created yet. Hence the sql constraint didn't check `date_from < date_to` yet.
Thanks